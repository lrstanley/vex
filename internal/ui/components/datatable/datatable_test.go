// Copyright (c) Liam Stanley <liam@liam.sh>. All rights reserved. Use of
// this source code is governed by the MIT license that can be found in
// the LICENSE file.

package datatable

import (
	"os"
	"testing"

	"github.com/gkampitakis/go-snaps/snaps"
	"github.com/lrstanley/vex/internal/api"
	"github.com/lrstanley/vex/internal/ui/state"
	"github.com/lrstanley/vex/internal/ui/testui"
)

func TestMain(m *testing.M) {
	v := m.Run()
	snaps.Clean(m, snaps.CleanOpts{Sort: true}) //nolint:errcheck
	os.Exit(v)
}

func TestNew(t *testing.T) {
	t.Parallel()
	t.Run("basic-table", func(t *testing.T) {
		t.Parallel()
		app := state.NewMockAppState(api.NewMockClient(), nil)
		m := New(app, Config[string]{
			NoResultsMsg: "no items found",
			RowFn: func(item string) []string {
				return []string{item, "description"}
			},
		})
		m.SetDimensions(testui.DefaultTermWidth, testui.DefaultTermHeight)
		tm := testui.NewNonRootModel(t, m, false)
		m.SetData([]string{"Name", "Description"}, []string{"item1", "item2"})
		tm.ExpectViewContains(t, "item1", "item2", "Name", "Description")
		tm.ExpectViewDimensions(t, m.GetWidth(), m.GetHeight())
		tm.ExpectViewSnapshot(t)
	})

	t.Run("loading-state", func(t *testing.T) {
		t.Parallel()
		app := state.NewMockAppState(api.NewMockClient(), nil)
		m := New(app, Config[string]{
			RowFn: func(item string) []string {
				return []string{item}
			},
		})
		m.SetDimensions(testui.DefaultTermWidth, testui.DefaultTermHeight)
		tm := testui.NewNonRootModel(t, m, false)
		tm.ExpectViewContains(t, "loading")
		tm.ExpectViewSnapshot(t)
	})

	t.Run("no-results", func(t *testing.T) {
		t.Parallel()
		app := state.NewMockAppState(api.NewMockClient(), nil)
		m := New(app, Config[string]{
			NoResultsMsg: "no items found",
			RowFn: func(item string) []string {
				return []string{item}
			},
		})
		m.SetDimensions(testui.DefaultTermWidth, testui.DefaultTermHeight)
		tm := testui.NewNonRootModel(t, m, false)
		m.SetData([]string{"Name"}, []string{})
		tm.ExpectViewContains(t, "no items found")
		tm.ExpectViewSnapshot(t)
	})

	t.Run("filtered-no-results", func(t *testing.T) {
		t.Parallel()
		app := state.NewMockAppState(api.NewMockClient(), nil)
		m := New(app, Config[string]{
			NoResultsFilterMsg: "no results for %q",
			RowFn: func(item string) []string {
				return []string{item}
			},
		})
		m.SetDimensions(testui.DefaultTermWidth, testui.DefaultTermHeight)
		tm := testui.NewNonRootModel(t, m, false)
		m.SetData([]string{"Name"}, []string{"item1", "item2"})
		m.SetFilter("nonexistent")
		tm.ExpectViewContains(t, "no results for \"nonexistent\"")
		tm.ExpectViewSnapshot(t)
	})

	t.Run("0-width-height", func(t *testing.T) {
		t.Parallel()
		app := state.NewMockAppState(api.NewMockClient(), nil)
		m := New(app, Config[string]{
			RowFn: func(item string) []string {
				return []string{item}
			},
		})
		m.SetDimensions(0, 0)
		tm := testui.NewNonRootModel(t, m, false, testui.WithTermSize(0, 0))
		m.SetData([]string{"Name"}, []string{"item1"})
		tm.ExpectViewSnapshot(t)
	})
}
