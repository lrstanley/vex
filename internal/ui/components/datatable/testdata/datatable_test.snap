
[TestNew/0-width-height - 1]

---

[TestNew/basic-table - 1]
[m[m [1mName [m  [1mDescription                                                            [m [m[m
[m[m[1m item1  description                                                             [m[m[m
[m[m item2  description                                                             [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
[m[m                                                                                [m[m
---

[TestNew/filtered-no-results - 1]
[m[m                         [m[m [m[mno results for "nonexistent"[m[m [m[m                         [m[m
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
---

[TestNew/loading-state - 1]
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                 [m[m<spinner>[m [mloading...[m [m                                  [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
[m                                                                                [m
---

[TestNew/no-results - 1]
[m[m                                [m[m [m[mno items found[m[m [m[m                                [m[m
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
                                                                                
---
